storage: add information about the size of the freed keys/values to GCInfo

dsaveliev opened the pull request:
Related to #20458

```bash
$ ./cockroach debug estimate-gc ./cockroach-data
...
RangeID: 4 [/System/NodeLivenessMax, /System/tsd):
storage.GCInfo{
    Now:                        hlc.Timestamp{WallTime:1512590365477252547, Logical:0},
    Policy:                     config.GCPolicy{TTLSeconds:86400},
    NumKeysAffected:            4,
    IntentsConsidered:          0,
    IntentTxns:                 0,
    TransactionSpanTotal:       0,
    TransactionSpanGCAborted:   0,
    TransactionSpanGCCommitted: 0,
    TransactionSpanGCPending:   0,
    TxnSpanGCThreshold:         hlc.Timestamp{WallTime:1512586765477252547, Logical:0},
    AbortSpanTotal:             0,
    AbortSpanConsidered:        0,
    AbortSpanGCNum:             0,
    PushTxn:                    0,
    ResolveTotal:               0,
    ResolveSuccess:             0,
    Threshold:                  hlc.Timestamp{WallTime:1512503965477252547, Logical:0},
    AffectedVersionsKeyBytes:   367,
    AffectedVersionsValBytes:   213,
}
...
```

* About logging - **RunGC** has [this](https://github.com/cockroachdb/cockroach/blob/2934e987f0c7c12c1cd374cc741910f9532f9d9b/pkg/storage/gc_queue.go#L904) for **GCInfo**. Do we need an additional logging of new fields?
* Do I need to print out an info about the total space that we could free at the end of output of `debug estimate-gc `?

dsaveliev committed:
storage: report version key/value bytes deleted in GCInfo

Report AffectedVersionsKeyBytes and AffectedVersionsValBytes in GCInfo.

Fixes #20458.

Release note: None.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20532)
<!-- Reviewable:end -->

tschottdorf commented:
Thanks @dsaveliev, that looks good! Only a few comments.

Could you add a commit message? I'm thinking you could title your commit `storage: report version key/value bytes deleted in GCInfo` and message

> Report AffectedVersionsKeyBytes and AffectedVersionsValBytes in GCInfo.
>
> Fixes #20458.
>
> Release note: None.

Could you add a test for this? I think it's straightforward to adapt `TestGCQueueProcess`:

1. before running the GC (before calling `gcQ.processImpl`), you manually invoke `RunGC` with dummy functions that don't actually delete anything and check the returned `GCInfo`
2. It's a bit tedious to actually compute the real bytes, but I think it's a good exercise to write down the expression. I think for the values, you end up with pretty much `<number_of_values_GCed>*<some constant larger than len("value")>` and similarly for keys (potentially not quite as nicely since there are differences in how meta keys vs non-meta keys are encoded, etc). See how far you get (and make sure you add commentary on your computation).

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/storage/gc_queue.go, line 676 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532#-L-hY79TVnJ0i_7qtnLZ:-L-hY79Ujksl5mPtlOyW:b-c0c81f) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c870b83502633449c0aab4ab1d8ba387f5cbeff/pkg/storage/gc_queue.go#L676)):*
> ```Go
> 	// Threshold is the computed expiration timestamp. Equal to `Now - Policy`.
> 	Threshold hlc.Timestamp
> 	// AffectedVersionsKeyBytes is the total size of the freed keys versions.
> ```

// AffectedVersionsKeyBytes is the number of (fully encoded) bytes deleted from keys in the storage engine.
// Note that this does not account for compression that the storage engine uses to store data on disk. Real
// space savings tends to be smaller due to this compression, and space may be released only at a later point
// in time.

---

*[pkg/storage/gc_queue.go, line 678 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532#-L-hZJKbsYQE-HUfsjI1:-L-hZJKbsYQE-HUfsjI2:b-o81r7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c870b83502633449c0aab4ab1d8ba387f5cbeff/pkg/storage/gc_queue.go#L678)):*
> ```Go
> 	// AffectedVersionsKeyBytes is the total size of the freed keys versions.
> 	AffectedVersionsKeyBytes int64
> 	// AffectedVersionsKeyBytes is the total size of the freed values versions.
> ```

// AffectedVersionsValBytes is the number of (fully encoded) bytes deleted from values in the storage engine.
// See AffectedVersionsKeyBytes for caveats.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532#-:-L-hZnhEtL883bmbb56S:b-p03xfk)*
<!-- Sent from Reviewable.io -->

dsaveliev commented:
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/gc_queue.go, line 676 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532#-L-hY79TVnJ0i_7qtnLZ:-L0GpuukyLRUBBA7rl9f:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddeeeacb00e791264536bdf8dc1095c9ccd685aa/pkg/storage/gc_queue.go#L676)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

// AffectedVersionsKeyBytes is the number of (fully encoded) bytes deleted from keys in the storage engine.
// Note that this does not account for compression that the storage engine uses to store data on disk. Real
// space savings tends to be smaller due to this compression, and space may be released only at a later point
// in time.
</blockquote></details>

Fixed

---

*[pkg/storage/gc_queue.go, line 678 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532#-L-hZJKbsYQE-HUfsjI1:-L0GpwElt-GQp8Q2BiiZ:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddeeeacb00e791264536bdf8dc1095c9ccd685aa/pkg/storage/gc_queue.go#L678)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

// AffectedVersionsValBytes is the number of (fully encoded) bytes deleted from values in the storage engine.
// See AffectedVersionsKeyBytes for caveats.
</blockquote></details>

Fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, thanks @dsaveliev! Much appreciated.

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 776 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532#-L0GrGTfb_tETTiL9KDj:-L0GrGTfb_tETTiL9KDk:be5l812) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43247277c2870b8f3a225c6e079bacc612c99f7/pkg/storage/gc_queue.go#L776)):*
> ```Go
> 						infoMu.GCInfo.AffectedVersionsValBytes += int64(len(vals[i]))
> 
> 						batchGCKeysBytes += int64(len(keys[i].Key))
> ```

Actually this should use `EncodedSize()` too. :smile: 

---

*[pkg/storage/gc_queue_test.go, line 490 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532#-L0GrTSf7SUkRYlc32hS:-L0GrTSf7SUkRYlc32hT:b-ceilom) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43247277c2870b8f3a225c6e079bacc612c99f7/pkg/storage/gc_queue_test.go#L490)):*
> ```Go
> 	// key5 at ts1  (14 bytes) => "value" (10 bytes)
> 	// key5 at ts2  (14 bytes) => delete (0 bytes)
> 	// key10 at ts1 (14 bytes) => detele (0 bytes)
> ```

`delete`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532#-:-L0Gr_Ggzrg7eJNwE4Qn:b-1ax6ob)*
<!-- Sent from Reviewable.io -->

dsaveliev commented:
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/gc_queue.go, line 776 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532#-L0GrGTfb_tETTiL9KDj:-L0H0xLsJrPZ9wzYfCq5:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43247277c2870b8f3a225c6e079bacc612c99f7/pkg/storage/gc_queue.go#L776)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Actually this should use `EncodedSize()` too. :smile: 
</blockquote></details>

Done

---

*[pkg/storage/gc_queue_test.go, line 490 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532#-L0GrTSf7SUkRYlc32hS:-L0H0uVYqSC1A16EyGdp:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43247277c2870b8f3a225c6e079bacc612c99f7/pkg/storage/gc_queue_test.go#L490)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

`delete`
</blockquote></details>

Fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/gc_queue_test.go, line 984 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532#-L0HdOZuapT0hcD_WfKR:-L0HdOZvIX6vHHGRGHW5:breidt6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d012ae0f53194f0e9e6a088bb197e03407fb615f/pkg/storage/gc_queue_test.go#L984)):*
> ```Go
> 		t.Fatalf("expected gcKeyVersionChunkBytes to be a multiple of %d", keyCount)
> 	}
> 	const keySize = gcKeyVersionChunkBytes / keyCount
> ```

Subtract the length of the timestamp suffix here. (the difference between `len(keys[i].Key)` and the `EncodedSize()` one).

---

*[pkg/storage/gc_queue_test.go, line 1049 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532#-L0HdbajU_9UYc1mh0wi:-L0HdbajU_9UYc1mh0wj:b-5hp44f) ([raw file](https://github.com/cockroachdb/cockroach/blob/d012ae0f53194f0e9e6a088bb197e03407fb615f/pkg/storage/gc_queue_test.go#L1049)):*
> ```Go
> 	// five batches. There is also a first batch which sets the GC
> 	// thresholds, making six total batches.
> 	if a, e := atomic.LoadInt32(&gcRequests), int32(7); a != e {
> ```

Then this can be changed back (the test doesn't make sense like this).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532)*
<!-- Sent from Reviewable.io -->

dsaveliev commented:
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/gc_queue_test.go, line 984 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532#-L0HdOZuapT0hcD_WfKR:-L0L_vWpjL5r1Im6toBC:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d012ae0f53194f0e9e6a088bb197e03407fb615f/pkg/storage/gc_queue_test.go#L984)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Subtract the length of the timestamp suffix here. (the difference between `len(keys[i].Key)` and the `EncodedSize()` one).
</blockquote></details>

Fixed

---

*[pkg/storage/gc_queue_test.go, line 1049 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532#-L0HdbajU_9UYc1mh0wi:-L0L_sgRzo3CIF9GBfmo:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d012ae0f53194f0e9e6a088bb197e03407fb615f/pkg/storage/gc_queue_test.go#L1049)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Then this can be changed back (the test doesn't make sense like this).
</blockquote></details>

Fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Thanks! Triggered another CI run, you seem to have caught an unrelated flake. I'll merge this when it's green.

tschottdorf commented:
Reviewed 1 of 1 files at r4.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20532)*
<!-- Sent from Reviewable.io -->
